### ISSUE TITLE
GitHub: Specially handle 502 response for graphql

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#3862]
GitHub's GraphQL backend tends to barf if it takes too long to process a query. I *think* that it may return a 502 response in some cases, which I also *think* gets automatically retried by `got`.

We should:
- Disable 502 retries for graphql POSTs, if we confirm that's happening
- Use the 502 response as a trigger to reduce our requests per-page

**Comment by rarkins** [comment]
Example failure:

```json
{
  "hostname": "ip-172-31-10-236",
  "level": 30,
  "repository": "corux\/alexa-flagquiz-skill",
  "err": {
    "name": "HTTPError",
    "hostname": "api.github.com",
    "method": "POST",
    "path": "\/graphql",
    "protocol": "https:",
    "url": "https:\/\/api.github.com\/graphql",
    "gotOptions": {
      "path": "\/graphql",
      "protocol": "https:",
      "hostname": "api.github.com",
      "hash": "",
      "search": "",
      "pathname": "\/graphql",
      "href": "https:\/\/api.github.com\/graphql",
      "headers": {
        "user-agent": "Renovate Bot (GitHub App 2740)",
        "accept": "application\/vnd.github.machine-man-preview+json, application\/vnd.github.merge-info-preview+json",
        "accept-encoding": "gzip, deflate",
        "authorization": "** redacted **",
        "content-length": 418
      },
      "hooks": {
        "beforeError": [
          
        ],
        "init": [
          
        ],
        "beforeRequest": [
          
        ],
        "beforeRedirect": [
          
        ],
        "beforeRetry": [
          
        ],
        "afterResponse": [
          
        ]
      },
      "retry": {
        "methods": {
          
        },
        "statusCodes": {
          
        },
        "errorCodes": {
          
        }
      },
      "decompress": true,
      "throwHttpErrors": true,
      "followRedirect": true,
      "stream": false,
      "form": false,
      "json": false,
      "cache": false,
      "useElectronNet": false,
      "hostType": "github",
      "baseUrl": "https:\/\/api.github.com\/",
      "body": "{\"query\":\"\\n  query {\\n    repository(owner: \\\"corux\\\", name: \\\"alexa-flagquiz-skill\\\") {\\n      issues(first: 100, after:null, orderBy: {field: UPDATED_AT, direction: DESC}, filterBy: {createdBy: \\\"renovate[bot]\\\"}) {\\n        pageInfo {\\n          startCursor\\n          hasNextPage\\n        }\\n        nodes {\\n          number\\n          state\\n          title\\n          body\\n        }\\n      }\\n    }\\n  }\\n  \"}",
      "method": "POST"
    },
    "statusCode": 502,
    "statusMessage": "Bad Gateway",
    "headers": {
      "date": "Sun, 02 Jun 2019 08:41:02 GMT",
      "content-type": "application\/json",
      "transfer-encoding": "chunked",
      "connection": "close",
      "server": "GitHub.com",
      "x-github-request-id": "D4A8:1638:141623D:184B05E:5CF38B9B"
    },
    "body": "{\n   \"data\":\"null\",\n   \"errors\":[\n      {\n         \"message\":\"Something went wrong while executing your query. This may be the result of a timeout, or it could be a GitHub bug. Please include `D4A8:1638:141623D:184B05E:5CF38B9B` when reporting this issue.\"\n      }\n   ]\n}\n",
    "message": "Response code 502 (Bad Gateway)",
    "stack": "HTTPError: Response code 502 (Bad Gateway)\n    at EventEmitter.emitter.on (\/home\/ubuntu\/renovateapp\/node_modules\/got\/source\/as-promise.js:74:19)\n    at process._tickCallback (internal\/process\/next_tick.js:68:7)"
  },
  "msg": "GitHub failure: 5xx",
  "time": "2019-06-02T08:41:02.256Z"
}
```

**Comment by Yanran1123** [comment]
Reducing our requests per page is very helpful. Thank you for the suggestion!

**Comment by olegkrivtsov** [comment]
I'd like to take this one.

**Comment by rarkins** [comment]
@olegkrivtsov this might already be implemented as part of the changes made by @zharinov recently. Please check it out and close this if so. The logic for reducing query size has certainly been added already

**Comment by olegkrivtsov** [comment]
Hi @rarkins I was able to find a couple of files mentioning GraphQL:

1. https://github.com/renovatebot/renovate/blob/eeedadede767ecd3c9728f67c672f14a382c84bf/lib/util/http/github.ts#L260
2. https://github.com/renovatebot/renovate/blob/main/lib/platform/github/index.ts

I don't see neither the special handling of the 502 error code nor the query size reducing logic. Could it be that the recent PR you are mentioning was not merged yet?

**Comment by rarkins** [comment]
Here is where we reduce the pagination size: https://github.com/renovatebot/renovate/blob/a9d334874aa172ba9afc7ffdc526b405c2f43411/lib/util/http/github.ts#L348

**Comment by olegkrivtsov** [comment]
Thank you. I think we don't handle the 502 error separately. Here is how the errors between 500 and 600 are handled:

https://github.com/renovatebot/renovate/blob/a9d334874aa172ba9afc7ffdc526b405c2f43411/lib/util/http/github.ts#L71

My plan would be to add another else-if and check for code 502. It it equals, throw a special exception so it stops iterating.

**Comment by rarkins** [comment]
Don't we want to continue iterating, but with a lower pagination count?

**Comment by olegkrivtsov** [comment]
Regarding POST requests which result in error 502, got doesn't retry them by default: https://github.com/sindresorhus/got/blob/main/documentation/7-retry.md#methods

"By default, Got does not retry on POST."

I don't see any places in the code that would override this behavior.

I researched the `graphql` code to understand how exactly it handles error 502, and it seems that it already handles it correctly - reduces the page size twice and continues iterating. I created this PR https://github.com/renovatebot/renovate/pull/13167 and added a unit test in case you would like to ensure error 502 is handled properly.
